/**
 */
package org.nasdanika.models.meta.impl;

import org.eclipse.emf.ecore.EClass;

import org.nasdanika.models.meta.Feature;
import org.nasdanika.models.meta.MetaPackage;

/**
 * <!-- begin-user-doc -->
 * An implementation of the model object '<em><b>Feature</b></em>'.
 * <!-- end-user-doc -->
 * <p>
 * The following features are implemented:
 * </p>
 * <ul>
 *   <li>{@link org.nasdanika.models.meta.impl.FeatureImpl#isDerived <em>Derived</em>}</li>
 *   <li>{@link org.nasdanika.models.meta.impl.FeatureImpl#isReadonly <em>Readonly</em>}</li>
 *   <li>{@link org.nasdanika.models.meta.impl.FeatureImpl#getDefaultValue <em>Default Value</em>}</li>
 * </ul>
 *
 * @generated
 */
public abstract class FeatureImpl extends TypedElementImpl implements Feature {
	/**
	 * The default value of the '{@link #isDerived() <em>Derived</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @see #isDerived()
	 * @generated
	 * @ordered
	 */
	protected static final boolean DERIVED_EDEFAULT = false;

	/**
	 * The default value of the '{@link #isReadonly() <em>Readonly</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @see #isReadonly()
	 * @generated
	 * @ordered
	 */
	protected static final boolean READONLY_EDEFAULT = false;

	/**
	 * The default value of the '{@link #getDefaultValue() <em>Default Value</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @see #getDefaultValue()
	 * @generated
	 * @ordered
	 */
	protected static final String DEFAULT_VALUE_EDEFAULT = null;

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	protected FeatureImpl() {
		super();
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	protected EClass eStaticClass() {
		return MetaPackage.Literals.FEATURE;
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	public boolean isDerived() {
		return (Boolean)eDynamicGet(MetaPackage.FEATURE__DERIVED, MetaPackage.Literals.FEATURE__DERIVED, true, true);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	public void setDerived(boolean newDerived) {
		eDynamicSet(MetaPackage.FEATURE__DERIVED, MetaPackage.Literals.FEATURE__DERIVED, newDerived);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	public boolean isReadonly() {
		return (Boolean)eDynamicGet(MetaPackage.FEATURE__READONLY, MetaPackage.Literals.FEATURE__READONLY, true, true);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	public void setReadonly(boolean newReadonly) {
		eDynamicSet(MetaPackage.FEATURE__READONLY, MetaPackage.Literals.FEATURE__READONLY, newReadonly);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	public String getDefaultValue() {
		return (String)eDynamicGet(MetaPackage.FEATURE__DEFAULT_VALUE, MetaPackage.Literals.FEATURE__DEFAULT_VALUE, true, true);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	public void setDefaultValue(String newDefaultValue) {
		eDynamicSet(MetaPackage.FEATURE__DEFAULT_VALUE, MetaPackage.Literals.FEATURE__DEFAULT_VALUE, newDefaultValue);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
			case MetaPackage.FEATURE__DERIVED:
				return isDerived();
			case MetaPackage.FEATURE__READONLY:
				return isReadonly();
			case MetaPackage.FEATURE__DEFAULT_VALUE:
				return getDefaultValue();
		}
		return super.eGet(featureID, resolve, coreType);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
			case MetaPackage.FEATURE__DERIVED:
				setDerived((Boolean)newValue);
				return;
			case MetaPackage.FEATURE__READONLY:
				setReadonly((Boolean)newValue);
				return;
			case MetaPackage.FEATURE__DEFAULT_VALUE:
				setDefaultValue((String)newValue);
				return;
		}
		super.eSet(featureID, newValue);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public void eUnset(int featureID) {
		switch (featureID) {
			case MetaPackage.FEATURE__DERIVED:
				setDerived(DERIVED_EDEFAULT);
				return;
			case MetaPackage.FEATURE__READONLY:
				setReadonly(READONLY_EDEFAULT);
				return;
			case MetaPackage.FEATURE__DEFAULT_VALUE:
				setDefaultValue(DEFAULT_VALUE_EDEFAULT);
				return;
		}
		super.eUnset(featureID);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
			case MetaPackage.FEATURE__DERIVED:
				return isDerived() != DERIVED_EDEFAULT;
			case MetaPackage.FEATURE__READONLY:
				return isReadonly() != READONLY_EDEFAULT;
			case MetaPackage.FEATURE__DEFAULT_VALUE:
				return DEFAULT_VALUE_EDEFAULT == null ? getDefaultValue() != null : !DEFAULT_VALUE_EDEFAULT.equals(getDefaultValue());
		}
		return super.eIsSet(featureID);
	}

} //FeatureImpl
